pegasus/helpers/hoc_helpers.rb

```diff
@@ -2,23 +2,69 @@
 require 'cdo/graphics/certificate_image'
 require 'dynamic_config/gatekeeper'
 
-def hoc_activity_writes_disabled
-  Gatekeeper.allows('hoc_activity_writes_disabled')
+UNSAMPLED_SESSION_ID = 'HOC_UNSAMPLED'
+
+# Creates a session row and sets the hour of code cookie to the session_id,
+# if the user is assigned to the sample set (as decided by a random choice
+# based on the hoc_activity_sample_proportion DCDO variable).
+#
+# If, however, the user is not in the sample, returns nil and sets the cookie
+# to UNSAMPLED_SESSION_ID.
+#
+# The "weight" encoded in the session row is set to 1/p, where p is the
+# proportion of sessions in the sample, so that reports can compute the
+# approximate number of actual sessions by summing over the the weights.
+
+def create_session_row_unless_unsampled(row)
+  # We don't need to do anything if we've already decided this session is unsampled.
+  return if unsampled_session?
+
+  # Decide whether the session should be sampled.
+  p = DCDO.get('hoc_activity_sample_proportion', 1.0).to_f
+  p = 1.0 if p == 0.0  # All sessions are in the sample if the proportion is invalid.
```

Comment:
We need to watch out for negative numbers, that's not meaningful.

I realize p = 0 *is* meaningful, that just means don't log.  There's no danger of divsion by zero since it's impossible to have rand() < p (because rand >= 0.)

The user might set the dcdo variable to a string but that will be converted to 0.0 by to_f.